### PR TITLE
Semantic analysis of expressions (work in progress)

### PR DESCRIPTION
**Pull request by klausler**
Here's a pile of new code that makes progress on the semantic analysis of expressions and the representation of typed expressions produced by successful analysis.

### CODE REVIEW

File: lib/common/kind-variant.h [hunk 1]
```diff
@@ -0,0 +1,121 @@
+// Copyright (c) 2018, NVIDIA CORPORATION.  All rights reserved.
```
Comment (klausler) [inline lib/common/kind-variant.h]: Oops!  I wrote this code and then ended up not using it anywhere.  Please ignore it until it disappears from this review.

File: lib/common/template.h [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+template<template<typename...> class T, typename... Ts>
+struct OverMembersHelper<T, std::variant<Ts...>> {
+  using type = T<Ts...>;
+};
+template<template<typename...> class T, typename... Ts>
+struct OverMembersHelper<T, std::tuple<Ts...>> {
+  using type = T<Ts...>;
+};
+
+template<template<typename...> class T, typename TorV>
```
Comment (tskeith) [inline lib/common/template.h]: It took me a while to realize that "TorV" meant "tuple or variant". Maybe that should be mentioned in a comment.

File: lib/common/template.h [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+  }
+  return std::nullopt;
+}
+
+// const std::variant<..., A, ...> -> std::optional<A>
+// i.e., when a variant holds a value of a particular type, return a copy
+// of that value in a std::optional<>.  The type A must be a valid
+// alternative for the variant.
+template<typename A, typename VARIANT>
+std::optional<A> GetIf(const VARIANT &u) {
```
Comment (tskeith) [inline lib/common/template.h]: What's the benefit of this over just using std::get_if? They seem almost interchangeable, but we have a lot of uses std::get_if already.

File: lib/common/template.h (hunk 2 above)
Comment (klausler) [reply lib/common/template.h]: I'll replace `TorV` with a longer name.

File: lib/common/template.h (hunk 3 above)
Comment (klausler) [reply lib/common/template.h]: I'll examine the uses of GetIf that survive and see whether they can be rewritten in terms of get_if.

File: lib/evaluate/type.h [hunk 4]
```diff
@@ -35,168 +38,245 @@ namespace Fortran::evaluate {
...
+
+  template<typename A> SomeScalar(const A &x) : u{x} {}
   template<typename A>
-  GenericScalar(std::enable_if_t<!std::is_reference_v<A>, A> &&x)
+  SomeScalar(std::enable_if_t<!std::is_reference_v<A>, A> &&x)
     : u{std::move(x)} {}
 
-  std::optional<std::int64_t> ToInt64() const;
-  std::optional<std::string> ToString() const;
+  auto ToInt64() const {
```
Comment (tskeith) [inline lib/evaluate/type.h]: I think these functions would be more readable with explicit return types.

File: lib/evaluate/type.h [hunk 5]
```diff
@@ -35,168 +38,245 @@ namespace Fortran::evaluate {
...
+          if constexpr (TypeOf<decltype(x)>::category ==
+              TypeCategory::Logical) {
+            return {x.IsTrue()};
+          }
+          return std::nullopt;
+        },
+        u);
+  }
+
+  template<typename T> auto GetIf() const {
```
Comment (tskeith) [inline lib/evaluate/type.h]: Especially this one, where the return type isn't even in the function body.

File: lib/evaluate/type.h (hunk 4 above)
Comment (klausler) [reply lib/evaluate/type.h]: OK.

File: lib/evaluate/type.h (hunk 5 above)
Comment (klausler) [reply lib/evaluate/type.h]: Deleting this function anyway after earlier review comment.

**Comment by klausler** [comment]
I've pushed an update that makes this branch current with the master branch and incorporates Tim's recent changes, so there's no longer any `dynamic_cast` usage.

**Pull request merged**